Comment Overhaul

iamrecursion opened the pull request:
In many programming languages, todo comments are simple to add, and can be parsed by external tools. In Luna, however, the dual-syntax representation means that they require special consideration. This proposal introduces useful todo comments to Luna that can be displayed in both the visual and textual syntaxes.

[Rendered](https://github.com/luna/luna-rfcs/blob/todo-comments/text/0000-comments-overhaul.md)

iamrecursion commented:
Shepherd is @kustosz!

wdanilo commented on text/0000-todo-comments.md:
```diff
@@ -0,0 +1,176 @@
+# Unresolved Questions
+
+- How do we ensure a _truly_ equivalent conversion between textual and visual
+  syntaxes?
```
This is VERY important question. The best answer Ive got so far is to always treat comments (doc/fixme/todo/etc) as something attached to the "line below" (or node representaed as the line below). This way we always know where we should display them and it introduces a great consistency - where to define comments. If you want to create todo comment to indicate that you want to create here code that do not exist yet, this rule tells you how to do it right! If you want to tell that there should be a function foo, but you don't provide its implementation yet, the following code is ok:

```
## TODO Let's implement it!
def foo
```

If you want to tell that a line (or several lines) are missing, do something like:

```
def foo:
   a = 17
   ## TODO Fix that mockup!
   b = a * 3
   ...
```

or 

```
def foo:
   a = 17
   ## TODO Add something here
   pass
   ...
```

We can use any other word for pass - this is just function returning Nothing.

iamrecursion replied:
Agreed on this point. It'll be reflected in the updated proposal.

iamrecursion commented:
In light of the increasing links between this proposal and the doc-comment one I'm working on, I'm going to roll both of the proposals into one!

iamrecursion commented on text/0000-todo-comments.md:
````diff
@@ -0,0 +1,178 @@
+
+```
+def myFunc a b:
+    # TODO: Frobnicate the arguments
````
> Do we plan to support other "TODO-like" keywords? (FIXME, BUG, etc?). Id so, we should change the title of this proposal.

This will be included in the revised RFC. It's a very good point and I don't often use them so I didn't think about it.

> I don't think TODO comments should share the syntax with disable comments.

Agreed. In the next iteration of the proposal I will be proposing `#` as the symbol for doc comments, FIXME, TODO, BUG, etc. I will also update the proposal to consider disable and freeze comments, and potential syntaxes for them as well.

iamrecursion commented on text/0000-todo-comments.md:
```diff
@@ -0,0 +1,178 @@
+are associated with their enclosing scope for conversion between Luna's dual
+syntaxes, and so often occur at the top of their scope. 
+
+If you type it wrong, for example using a variant of the word 'todo' that isn't
```
I will consider this when revising the proposal. It's a very good point that in the case of combining it with doc comments that we hit a 'clash'.

iamrecursion commented on text/0000-todo-comments.md:
```diff
@@ -0,0 +1,178 @@
+    # ToDo: Do the thing.
+      ^--^
+
+Unexpected token 'ToDo' found after '#'. Did you perhaps mean one of:
```
Agreed. Will be fixed.

iamrecursion commented on text/0000-todo-comments.md:
````diff
@@ -0,0 +1,178 @@
+where `comment-char` is defined as for existing Luna comments:
+
+```
+todo-comment = "#", ("todo" | "TODO" | "Todo"), [":"], {comment-char};
````
I agree on this. I feel we should standardise on all caps for each of `BUG`, `TODO` and `FIXME`, but I prefer having the colon there. Why do you prefer not having the colon?

wdanilo replied:
It's not a strong preference, but it's because of several small reasons:

1. It's seems like something not necessary - if we use colon in Luna it actually tells us where the argument list ends, here it just tells nothing and is just another thing to type. Of course it can look nicer with it, but it's also could be achieved by coloring the `# TODO` part other than its body. However this point is just personal preference and here I'm still completely ok to use colons.

2. The real problem is what happens if somebody don't type colon - what will happen then? Should we interpret this as doc comment / throw an error / accept it? Every solution is a bit awkward to me - interpreting as doc comment seems wrong, throwing error is a little strange and accepting it makes it optional, which we rather try to avoid everywhere we use it. 

Anyway, these points are rather not important - just thoughts why my initial reaction was to drop it.

iamrecursion replied:
I'm not particularly attached to the colon, though I've always written one given the choice. I would've said throw an error, but it's probably better to just get rid of it and avoid the problem entirely!

wdanilo replied:
I was also always writing a colon here! The question I've got in my head is if that was to just partition the thing into two parts? Probably yes. If so, making distinct colors would work the same way? I don't know yet, but I suspect so.

iamrecursion replied:
That is definitely how I used it, yes. The thing about colouring is that it won't always be present in every editor! Also, some people prefer to work without syntax highlighting.

wdanilo replied:
That's true. And here is my standard question to such problem - what is the percentage of such cases? Official Luna editors will always do it right. How many people prefer working on code without highlighting? Minority, personally I don't know anyone. So for me, its not a real case here. But, as I said earlier, both ways are ok for me :)

iamrecursion replied:
I know a lot of people that do it, and I have a few editors that I might use that don't currently have Luna syntax highlighting. Then again, I'm really not fussed.

iamrecursion commented:
I've made significant updates to the proposal. It now encompasses all comment types in Luna and I hope that I've addressed the concerns raised so far! 

@wdanilo @kustosz

kustosz commented:
A few things that pop up in my mind:

1. I like how we'd be swapping `#` and `##` with this. Looks nicer and seems more intuitive.
2. There's a possible pain with doc comments possibly containing the word `FIXME` at the beginning. Maybe the key-worded comments should not have a space between the `#` and the keyword? (like `#FIXME` instead of `# FIXME`).
3. I'm not sure about fixing the syntax for todos, fixmes etc. This is something that for me should be handled by convention, not by a fixed syntax. I mean, people have lots of different workflows and I wouldn't go as far as telling them what kinds of comments are OK, at least not without a way to extend it. So I feel there should be a syntax for "tagged" comments and our tooling should recognize the most common patterns, but going as far as restricting it on a syntax level sounds like on overkill to me.
4. Why not enrich markdown/reST with code references in the doc comments? I wouldn't want to learn yet another markup language just to be able to write documentation. What are the advantages of not reusing some standard?

iamrecursion commented:
> There's a possible pain with doc comments possibly containing...

This is actually handled in the proposal, but not made clear enough. Tagged comments are only parsed, merging lines, to the first blank comment line. This means that the following is parsed as a tagged comment AND doc comment:

```
# TODO Foo bar 
# another line that gets merged with the above 
#
# This is now part of the doc comment. 

Is that a satisfactory solution? 

> I'm not sure about fixing the syntax for todos, fixmes etc...

The notion behind this was to pursue the ‘one way to do things in Luna’ idea. I’m not particularly bothered one way or another as to whether we fix the syntax or not, but that’s the original rationale behind it. 

> Why not enrich markdown/reST with code references in the doc comments...

We certainly could! There’s even the potential for a preprocessor step to transform the additional syntax into valid markdown. I just think that the markdown spec is giant and most of it entirely irrelevant for doc comments, but that’s more my perspective than anything else.

wdanilo commented:
2. I don't see any problem with that. if somebody creates doc comment starting with `TODO` we can threat is as TODO comment. It would be counter intuitive for me if we allow for `#TODO something`, but `# TODO something` would give me a doc comment instead.

3. For me this discussion is exactly the same as language extensions. If we want to provide one good solution, let's tell people what syntax there is. Otherwise we can make extensible doc syntax and language syntax, but I'd really hate this idea.

4. reST is ugly. Markdown could be ok, however again, I hate that you can make the same thing in 100 different ways, like heading using `-` or `=` or `#` . Just no. Additionaly, the proposed syntax is much easier to read because it is hierarchical by using the indentation, just like Luna. For me it has big advantages over reST / Markdown - there is one way to write it and everyone will write it the same AND it's similar to Luna syntax.

kustosz commented:
Ad. 3. This is very different from language extensions. In fact, after some more consideration, I don't really get the point of calling fixmes syntax at all. I mean, at this point it seems like `# FIXME` will work as good as `# FOO`. The only difference is what we're supporting through tooling, nothing to do with language semantics (or parsing, for that matter) at all. Which also makes my previous point invalid, if I understand correctly there's nothing stopping me from using `# FOO` other than the fact `# FIXME` will get nicer formatting in some places, so that's fine.

Ad. 4. OK, I don't have that strong an opinion, it just seems kind of pointless from my perspective. But if there's someone willing to walk that extra mile to implement a whole-new markup language, let's go for it. It is a burden from maintenance perspective though, and one I'm not very keen on taking  considering how many more awesome things there are to implement for Luna.

kustosz commented:
OK, yet another thought about the `FIXME`s etc., and a clarification for what lead me astray when thinking about.
After some consideration – this is not at all another type of comment. It seems that it's just a keyword in the proposed markup language for doc comments. I mean, `FIXME Foo bar baz` is just something that generates `<p class="fixme">Foo bar baz</p>` inside the generated do comment.
I'd suggest we merge the separate sections for fixmes and todos into the doc-comment section, as it really is functionally just a part of it.
My initial reaction was triggered by the fact that I was still thinking about it as a part of language syntax – i.e. something that core luna parser would care about. Once that becomes a job for the markup processor, it's all fine :)

iamrecursion commented:
I like the idea of tagging comments becoming part of the doc comment syntax. I think that’s probably the best solution. 

The more I think about it, the less I want markdown as our doc comment syntax. I’d love to implement this, so I’d be happy to maintain our own custom syntax. Most languages have a custom syntax anyway.

Spaceghost commented:
This is rather relevant to the idea Literate Programming in Luna. I'd personally love to see a style where there's no code comments, but instead Luna would detect the Luna code amongst the prose.

Whether this could be the default or whether Luna wants to go a more traditional route, it would be exciting to have Literate Luna.

Regardless of how non-code is recognized in Luna, it would be wonderful if the comments were programmatically available in Luna for the visualizations and others to make use of.

👍

wdanilo commented:
@Spaceghost what is the story behind Literate Luna? What it gives to the user? To be completely honest here, I've been never using any literate style because I havent seen any benefit of using it. Moreover, it goes against the way to make writing Luna in an obvious way (now you'll have two ways to write the same). Basically it's not such a big problem for me, because we can distinguish this by file extensions and just assume that you either use Literate Luna or normal Luna, but I personally just don't see any benefits of supporting it.

wdanilo commented:
Moreover, according to https://en.wikipedia.org/wiki/Literate_programming : 

> Literate programming is a programming paradigm introduced by Donald Knuth in which a program is given as an explanation of the program logic in a natural language, such as English, interspersed with snippets of macros and traditional source code, from which a compilable source code can be generated. The literate programming paradigm, as conceived by Knuth, represents a move away from writing programs in the manner and order imposed by the computer, and instead enables programmers to develop programs in the order demanded by the logic and flow of their thoughts. Literate programs are written as an uninterrupted exposition of logic in an ordinary human language, much like the text of an essay, in which macros are included to hide abstractions and traditional source code.

Which basically makes sense if you think about languages which do not allow the reader to easily follow the logic. Using purely functional languages with good function names is readable and allows you to read sources the same way as you would read explanation what code actually does. This way you enforce better code quality and you are sure that code and its description will not sync out. Thus I believe that Luna do not need literate mode. If you're thinking more about something like Python notebook, then yes, something like that should exist as one of Luna Studio tools, but for me, it's not a lang-mod per se.

iamrecursion commented:
I don't think there's a lot of benefit to Literate Luna. I think much of the same effect can be realised by a notebook-style mode of operation, which I'm planning support for as part of #5.

wdanilo commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+Luna actually provides _two_ ways to do this, the first of which is known as a
+'disable comment'. Disable comments in Luna are very similar to commenting out
+code in any other language. Just prefix a line with `##` and it'll stop the code
+from executing. If you're in the graph view, select some nodes, and hit `Ctrl+/`
```
I would definitely consider making disabled nodes work like Houdini ones. So after disabling a node it passes its inputs to outputs somehow - we need to discuss the syntax and be sure it is clear in textual language. In the visual language we could do this easily - just like houdini does. EDIT: Like Houdini used to do. I was tryying to do it in the current version and while everything works as it did before, there is no visualisation of data flowing trought the node.

Basically - we can make nodes a little "transparent" and show how input flows to output

iamrecursion replied:
I've included a version of this in the updated proposal to give us a springboard for further discussion. You're right in that it is very important.

iamrecursion commented:
A new version is up @wdanilo and @kustosz!

wdanilo commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+actually slightly different! Say you want to work on some code internally while 
+still preserving the correct evaluation of your program. Luna supports a comment
+called a 'freeze comment', that makes an expression (or set of expressions) 
+return the last calculated value. These look like a `#|`, and also prefix the
```
I'm convinced the option 2. is the right one here.

iamrecursion replied:
That would be my intention, but for editors that aren't Luna studio un-commenting and re-commenting will have to suffice, so this should have the same effect. I will clarify this in the text.

wdanilo commented on text/0000-comments-overhaul.md:
````diff
@@ -0,0 +1,384 @@
+
+```
+def main:
+    #| hello = "Hello, World!"
````
After thinking about it more, the freeze comment have only sense where there is a variable which we bake the output to. Thus it has a big sense to combine the operators. See also comment end the end of this review.

iamrecursion replied:
I think it should remain a comment, as it's an external modifier to the code behaviour, and you may only want things frozen for a short amount of time. To me it's far less onerous to change a comment than to actually adapt the code.

wdanilo replied:
The longer I think about it, the more I'm convinced it should rather be something like `b #= a.foo bar`. Because:

1. This comment does not have any meaning if there is no variable you bake your computations to.
2. It's much more clear what it actually does.
3. I don't know if I would consider it to be a "comment". It is comment like syntax, however in reality it just changes the meaning of equality operator - it's just - evaluate once and cache. This is not a "comment", however it has slightly connected behavior - you're disabling some code after it gets evaluated. But it's more a syntax than a comment to such degree, that it even might be a separate RFC.
4. It has a notion of disabling some code (some computation), thus I like the idea of reusing the symbol `#` to indicate that something is about to be disabled.

There are also other points of view here. Again, its not a real comment. Thus we could allow to freeze only some patterns, like:

```
[a,b, ||c||] = foo bar # pseudo syntax
```

Where `c` will be frozen. I would personally keep it simpler and allow for frozing the whole computation only, not parts of it, but this also shows that we're dealing here with a syntax affecting semantics more than just a comment!

iamrecursion replied:
If that's the case then it shouldn't be part of this RFC and instead part of a separate one.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+as quickly as possible. As always, this means that something gets left out, and
+you want to add something to document that fact so you can come back later.
+
+Luna features `TODO` comments for this purpose, allowing you to annotate lines
```
I'll clarify.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+
+There are a few things to keep in mind when writing your documentation comments:
+
+- The first line should be a short explanation of whatever you're documenting, 
```
It already is. The first line is denoted by a blank line by virtue of the syntax used.

wdanilo replied:
I'll describe it in more explicit way though, so people reading it will clearly see how these decisions affect the header.

iamrecursion replied:
Will do.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+# 
+# blank line.
+# 
+# !Block Heading
```
A typo haha.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+#       * Bulleted sub-list, but you can mix numbers and bullets.
+# 
+# Sometimes I might want to embed a code block in my documentation as a usage
+# example. I can do this as > frobnicate foo bar <, but for bigger blocks I can:
```
Backticks are definitely more concise.

wdanilo replied:
And what with multi line code then?

iamrecursion replied:
If we're going to use them for inline code then I think we should use them for block code as well.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,384 @@
+#   > import Math
+#   > frobnicate "a" (Math.sqrt b)
+# 
+# @a: Here I'm making an item reference to the first function argument.
```
Backticks here would mean we can't use them for inline code. Just to keep that in mind.

wdanilo replied:
Actually we could. If we agree that we try to discover every name in backticks and bind it to the API defined in such module, then it will work both ways: `a` will refer to local variable `a`, while if you'll write multiline code, all known names would also be clickable.

iamrecursion replied:
Sure, if that's what you'd like!

wdanilo replied:
If that doesnt break anything, its very cool solution for me because it makes the syntax much more obvious - when you refer to code, just use backticks. Thats it.

iamrecursion commented on text/0000-comments-overhaul.md:
```diff
@@ -0,0 +1,423 @@
+
+## Commenting Out Code
+You're working on a feature and something isn't going quite right. You've tried
+to isolate the problem using the graph's visual debugging, but that hasn't 
```
I'll clarify.
